### PR TITLE
Rewrite plugin Search to es6.

### PR DESCRIPTION
**Pull request by pnowak**
### Context
Search is now based on BasePlugin.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Related issue(s):
1. #4629

### Checklist:

- [x] My code follows the code style of this project,
- [x] My change requires a change to the documentation.

### CODE REVIEW

File: src/plugins/search/test/search.e2e.js [hunk 1]
```diff
@@ -347,66 +372,80 @@ describe('Search plugin', () => {
...
-
-          if (rowIndex == 1) {
-            expect($(cell).hasClass(Handsontable.plugins.Search.DEFAULT_SEARCH_RESULT_CLASS)).toBe(true);
-          } else {
-            expect($(cell).hasClass(Handsontable.plugins.Search.DEFAULT_SEARCH_RESULT_CLASS)).toBe(false);
-          }
-        }
-      }
+      let cell = hot.getCell(0, 0);
+      expect($(cell).hasClass(hot.getPlugin('search').searchResultClass)).toBe(false);
```
Comment (wszymanski) [inline src/plugins/search/test/search.e2e.js]: Please simplify this calls here and in next tests.

File: src/plugins/search/test/search.e2e.js [hunk 2]
```diff
@@ -135,183 +161,182 @@ describe('Search plugin', () => {
...
 
-      var searchResult = hot.search.query('A', searchCallback);
+      hot.getPlugin('search').query('A', searchCallback);
 
       expect(searchCallback.calls.count()).toEqual(25);
 
-      for (var rowIndex = 0, rowCount = countRows(); rowIndex < rowCount; rowIndex++) {
-        for (var colIndex = 0, colCount = countCols(); colIndex < colCount; colIndex++) {
-          var callArgs = searchCallback.calls.argsFor((rowIndex * 5) + colIndex);
+      for (let rowIndex = 0, rowCount = countRows(); rowIndex < rowCount; rowIndex += 1) {
```
Comment (wszymanski) [inline src/plugins/search/test/search.e2e.js]: Please test it without looping.

File: src/plugins/search/search.js [hunk 3]
```diff
@@ -41,95 +158,135 @@ function Search(instance) {
...
+   * @param {String} value Value of the rendered cell.
+   * @param {Object} cellProperties Object containing the cell's properties.
+   */
+  onBeforeRenderer(TD, row, col, prop, value, cellProperties) {
+    let classArray = [];
 
-function SearchCellDecorator(instance, TD, row, col, prop, value, cellProperties) {
-  var searchResultClass = (cellProperties.search !== null && typeof cellProperties.search == 'object' &&
-      cellProperties.search.searchResultClass) || Search.global.getDefaultSearchResultClass();
+    if (cellProperties.className) {
```
Comment (wszymanski) [inline src/plugins/search/search.js]: Please add tests. Please cache the `cellProperties.className`.

File: src/plugins/search/search.js [hunk 4]
```diff
@@ -1,35 +1,152 @@
...
+    super.enablePlugin();
+  }
+
+  /**
+   * Disable plugin for this Handsontable instance.
+   */
+  disablePlugin() {
+    const beforeRendererCallback = (TD, row, col, prop, value, cellProperties) => this.onBeforeRenderer(TD, row, col, prop, value, cellProperties);
+
+    this.hot.addHook('beforeRenderer', beforeRendererCallback);
```
Comment (wszymanski) [inline src/plugins/search/search.js]: Please add tests.

File: src/plugins/search/search.js [hunk 5]
```diff
@@ -1,35 +1,152 @@
...
+     */
+    this.callback = DEFAULT_CALLBACK;
+    /**
+     * Query function is responsible for determining whether a query matches the value stored in a cell.
+     *
+     * @type {Function}
+     */
+    this.queryMethod = DEFAULT_QUERY_METHOD;
+    /**
+     *  Class added to each cell that belongs to the search.
```
Comment (wszymanski) [inline src/plugins/search/search.js]: Maybe change to: `Class added to each cell that belongs to the searched query.`

File: src/plugins/search/search.js [hunk 6]
```diff
@@ -1,35 +1,152 @@
...
-    var queryResult = [];
-
-    if (!callback) {
-      callback = Search.global.getDefaultCallback();
-    }
+class Search extends BasePlugin {
+  constructor(hotInstance) {
+    super(hotInstance);
+    /**
+     * Callback function is responsible for marking the cells belonging to the search by setting property to it.
```
Comment (wszymanski) [inline src/plugins/search/search.js]: Please describe the property, not the value.

File: src/plugins/search/search.js [hunk 7]
```diff
@@ -1,35 +1,152 @@
...
+    this.enablePlugin();
+
+    super.updatePlugin();
+  }
+
+  /**
+   * Query method - used inside search input listener.
+   *
+   * @param {String} queryStr Searched value.
+   * @param {Function} [callback=DEFAULT_CALLBACK] Callback function responsible for setting the cell's `isSearchResult` property.
```
Comment (wszymanski) [inline src/plugins/search/search.js]: Please change from `[callback=DEFAULT_CALLBACK]` to `[callback]`. Apply the same rule below.

File: src/plugins/search/search.js [hunk 8]
```diff
@@ -1,35 +1,149 @@
...
-      for (var colIndex = 0; colIndex < colCount; colIndex++) {
-        var cellData = instance.getDataAtCell(rowIndex, colIndex);
-        var cellProperties = instance.getCellMeta(rowIndex, colIndex);
-        var cellCallback = cellProperties.search.callback || callback;
-        var cellQueryMethod = cellProperties.search.queryMethod || queryMethod;
-        var testResult = cellQueryMethod(queryStr, cellData);
+    const searchSettings = this.hot.getSettings().search;
+    this.checkPluginSettings(searchSettings);
+
+    this.addHook('beforeRenderer', (TD, row, col, prop, value, cellProperties) => this.onBeforeRenderer(TD, row, col, prop, value, cellProperties));
```
[the remark this answers is not in the document]
Comment (pnowak) [reply src/plugins/search/search.js]: Done.

File: src/plugins/search/search.js (hunk 5 above)
Comment (pnowak) [reply src/plugins/search/search.js]: Done.

File: src/plugins/search/search.js (hunk 7 above)
Comment (pnowak) [reply src/plugins/search/search.js]: Done.

File: src/plugins/search/test/search.e2e.js (hunk 1 above)
Comment (pnowak) [reply src/plugins/search/test/search.e2e.js]: Done.

File: src/plugins/search/search.js (hunk 3 above)
Comment (pnowak) [reply src/plugins/search/search.js]: Done.

**Pull request merged**